File: 56-merge-intervals/merge-intervals.cpp
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        vector<pair<int,int>> startEndArr;
        for(auto it: intervals){
            startEndArr.push_back({it[0], it[1]});
        }

        sort(startEndArr.begin(), startEndArr.end());

        auto [prevStart, prevEnd] = startEndArr[0];
        vector<vector<int>> ans;

        for(int i=1;i<startEndArr.size();i++){
            // 8 ---- 10
            //            15 ------ 18
            auto [currStart, currEnd] = startEndArr[i];

            if(prevEnd >= currStart){
                // we can merger currMeeting is start before previous unterval end 
                prevEnd = max(prevEnd, currEnd);
            }else {
                ans.push_back({prevStart, prevEnd}); // {1,6}, {8,10}
                prevStart = currStart; // 15
                prevEnd = currEnd; // 16
            }
        }
        ans.push_back({prevStart, prevEnd});

        return ans;
        
    }
};

File: 56-merge-intervals/merge-intervals.java
class Solution {
    public int[][] merge(int[][] intervals) {
        List<int[]> list = new ArrayList<>();

        Arrays.sort(intervals, (a,b) -> Integer.compare(a[0], b[0]));

        int start = intervals[0][0];
        int end = intervals[0][1];

        for(int[] arr : intervals) {
            int currStart = arr[0];
            int currEnd = arr[1];

            if(currStart <= end){
                // can merge 
                end = Math.max(end, currEnd);
            }else{
                list.add(new int[] { start, end});
                start = currStart;
                end = currEnd;
            }

        }
        list.add(new int[] { start, end});
        return list.toArray(new int[list.size()][2]);
    }
}